File: mini-services/ai-service/src/cache.ts
// ═══════════════════════════════════════════════════════════════════════════
// Orbit — Micro-service IA · cache mémoire des réponses identiques
// ───────────────────────────────────────────────────────────────────────────
// Map ordonnée (FIFO par insertion) : 200 entrées max, TTL 10 min.
// Les requêtes /chat ne passent PAS par ce cache (réponses uniques par
// nature). Un compteur de hits alimente /health (observabilité).
// ═══════════════════════════════════════════════════════════════════════════

import { CACHE_MAX_ENTRIES, CACHE_TTL_MS } from "./config"

interface Entry<T> {
  value: T
  expiresAt: number
}

const store = new Map<string, Entry<unknown>>()
let hits = 0

export function cacheGet<T>(key: string): T | null {
  const entry = store.get(key)
  if (!entry) return null
  if (entry.expiresAt <= Date.now()) {
    store.delete(key)
    return null
  }
  hits++
  return entry.value as T
}

export function cacheSet(key: string, value: unknown): void {
  // Éviction du plus ancien élément au-delà de la capacité (Map = ordre d'insertion).
  if (store.size >= CACHE_MAX_ENTRIES) {
    const oldest = store.keys().next().value
    if (oldest !== undefined) store.delete(oldest)
  }
  store.set(key, { value, expiresAt: Date.now() + CACHE_TTL_MS })
}

export function cacheStats(): { size: number; hits: number } {
  // Purge opportuniste des entrées expirées pour un compteur honnête.
  const now = Date.now()
  for (const [k, e] of store) if (e.expiresAt <= now) store.delete(k)
  return { size: store.size, hits }
}

File: docker/ai-service/services/cache.py
"""Orbit — cache mémoire des réponses IA identiques.

Équivalent Python de src/cache.ts (mini-service bun) : dictionnaire ordonné,
200 entrées max, TTL 10 min, compteur de hits pour /health.
Les requêtes /chat ne passent pas par ce cache (réponses uniques par nature).
"""

import os
import time
from collections import OrderedDict

_TTL_S = float(os.getenv("AI_CACHE_TTL_MS", "600000")) / 1000.0
_MAX_ENTRIES = 200
_store: "OrderedDict[str, tuple]" = OrderedDict()
_hits = 0


def cache_get(key: str):
    entry = _store.get(key)
    if entry is None:
        return None
    value, expires_at = entry
    if expires_at <= time.time():
        _store.pop(key, None)
        return None
    global _hits
    _hits += 1
    return value


def cache_set(key: str, value) -> None:
    if len(_store) >= _MAX_ENTRIES:
        _store.popitem(last=False)  # éviction du plus ancien (FIFO)
    _store[key] = (value, time.time() + _TTL_S)


def cache_stats() -> dict:
    now = time.time()
    for key in [k for k, (_, exp) in _store.items() if exp <= now]:
        _store.pop(key, None)
    return {"size": len(_store), "hits": _hits}
